### PR TITLE
discocss: add module

### PR DESCRIPTION
**Pull request by Kranzes**
### Description

Rewrite of #3079

~~Waiting for https://github.com/NixOS/nixpkgs/pull/189164 to pass Hydra.~~

### Checklist

- [ ] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/programs/discocss.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+with lib;
+
+let cfg = config.programs.discocss;
+in {
+  meta.maintainers = with maintainers; [ kranzes ];
+
+  options = {
+    programs.discocss = {
+      enable = mkEnableOption
+        "discocss, atiny Discord css-injector for Linux and MacOS";
```
Comment (rycee) [inline modules/programs/discocss.nix:12]:
```suggestion
        "discocss, a tiny Discord CSS-injector for Linux and MacOS";
```

File: modules/programs/discocss.nix [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [{
+      assertion = cfg.discordAlias
+        -> !(builtins.any (p: p.name == cfg.discord.name) config.home.packages);
+      message =
+        "To use discocss with discordAlias you have to remove discord from home.packages, or set discordAlias = false;";
```
Comment (rycee) [inline modules/programs/discocss.nix:44]:
```suggestion
        "To use discocss with discordAlias you have to remove discord from home.packages, or set discordAlias = false.";
```

File: modules/programs/discocss.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        default = null;
+        description = "The custom CSS for discocss to use.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [{
+      assertion = cfg.discordAlias
+        -> !(builtins.any (p: p.name == cfg.discord.name) config.home.packages);
```
Comment (rycee) [inline modules/programs/discocss.nix:42]: I think it is sufficient to have

```suggestion
        -> !(any (p: p.name == cfg.discord.name) config.home.packages);
```

**Comment by rycee** [comment]
Thanks for the contribution! I've added a few comments.

File: modules/programs/discocss.nix (hunk 3 above)
Comment (Kranzes) [reply modules/programs/discocss.nix:42]: `any` isn't in scope, is it?

Comment (Kranzes) [reply modules/programs/discocss.nix:42]: oh it's coming from nixpkgs' lib.

**Comment by rycee** [comment]
Thanks! Merged to master now.

**Comment by eyadsibai** [comment]
I am getting 
```touch: cannot touch '/home/eyad/.config/discocss/custom.css': Permission denied```

**Comment by Kranzes** [comment]
Apparently the discocss version in nixpkgs is missing this commit, https://github.com/mlvzk/discocss/pull/17 which is needed for it to work. I'd suggest setting `programs.discocss,package` to `(builtins.getFlake "github:mlvzk/discocss/aa71249beed302d4b99bfaa16cda3e0929181f5f").packages.${pkgs.system}.discocss`

**Comment by Kranzes** [comment]
Another alternative is using this
```nix
programs.discocss = {
  enable = true;
  package = pkgs.discocss.overrideAttrs (_: {
    patches = [
      (pkgs.fetchpatch {
        url = "https://github.com/mlvzk/discocss/commit/83b53f3d08cd1d448caa4aa77a4a19f2fdc2f523.patch";
        sha256 = "sha256-iyKXJ7Xa9X1v0cyKnkCHmXpd1Iuy/VBuhRv4x8lz5Cs=";
      })
    ];
  });
  css = builtins.readFile "${pkgs.fetchurl {
    url = "https://raw.githubusercontent.com/orblazer/discord-nordic/8b62f70a6b61e63da1c5b603ff7f8205651e7d09/base.css";
    sha256 = "sha256-HMnGC5KmNYAU4j8lOdNzWPnm48z8YgkY59LxoCF9wB0="; }
  }";
};
```

**Comment by JavaCafe01** [comment]
> Apparently the discocss version in nixpkgs is missing this commit, [mlvzk/discocss#17](https://github.com/mlvzk/discocss/pull/17) which is needed for it to work. I'd suggest setting `programs.discocss.package` to `(builtins.getFlake "github:mlvzk/discocss/aa71249beed302d4b99bfaa16cda3e0929181f5f").packages.${pkgs.system}.discocss`. if @mlvzk doesn't merge the PR in the next couple of days I will just patch it in nixpkgs instead.

I have pinged @mlvzk on discord. Hopefully they respond, but if they don't I think the best option is patching it in nixpkgs like you said.

**Comment by JavaCafe01** [comment]
@rycee This module does not work at all, so I think this PR/issue should be reopened, or the patch should be implemented? I don't know what the convention is.

**Comment by eyadsibai** [comment]
btw this works for me... why would this module not work and need a patch?

{ config, pkgs, lib, ... }:
{
  home.packages = with pkgs; [ discord discocss ];

  xdg.configFile."discocss/custom.css".text = ''
    
  '';
}

**Comment by rycee** [comment]
If there is a problem then please open a new issue with a description. It would be good if the new issue references this PR.

FWIW if the upstream package need any patching then it should be done in Nixpkgs, not in Home Manager.

**Comment by Kranzes** [comment]
This will fix it https://github.com/NixOS/nixpkgs/pull/197788

**Comment by JavaCafe01** [comment]
@Kranzes https://github.com/mlvzk/discocss/pull/17 is merged btw :+1:

**Pull request merged**